Sources/RoktUXHelper/Data/Expansion/Extractor/BNFCreativeDataExtractor.swift

```diff
@@ -0,0 +1,136 @@
+//
+//  BNFCreativeDataExtractor.swift
+//  RoktUXHelper
+//
+//  Licensed under the Rokt Software Development Kit (SDK) Terms of Use
+//  Version 2.0 (the "License");
+//
+//  You may not use this file except in compliance with the License.
+//
+//  You may obtain a copy of the License at https://rokt.com/sdk-license-2-0/
+
+import Foundation
+
+@available(iOS 13, *)
+class BNFCreativeDataExtractor<Validator: DataValidator>: DataExtractor where Validator.T == String {
+
+    private let dataValidator: Validator
+    private let parser: PropertyChainDataParser
+    private let dataReflector: any DataReflector
+
+    init(
+        dataValidator: Validator = BNFPlaceholderValidator(),
+        parser: PropertyChainDataParser = BNFPropertyChainDataParser(),
+        dataReflector: any DataReflector = BNFDataReflector()
+    ) {
+        self.dataValidator = dataValidator
+        self.parser = parser
+        self.dataReflector = dataReflector
+    }
+
+    // maps `propertyChain` to a value inside `Offer`, if possible
+    // note that you can chain statements like A? | B? | Default?
+    // if A does not exist, check B, else use default
+    func extractDataRepresentedBy<U>(
+        _ type: U.Type,
+        propertyChain: String,
+        responseKey: String?,
+        from data: OfferModel?
+    ) throws -> DataBinding<U> {
+        guard dataValidator.isValid(data: propertyChain) else { return .value(propertyChain as! U) }
+
+        let placeholder = parser.parse(propertyChain: propertyChain)
+
+        var isStateType = false
+
+        var mappedData: String?
+
+        for keyAndNamespace in placeholder.parseableChains {
```

Comment:
For this and other classes this could likely be broken out into a function to make parsing easier. Ultimately this for loop reads as `getMappedDataFromParsableChains()`

Reply:
agreed, could improve this in another PR, this file was renamed from BNFDataExtractor.